psi4/driver/p4util/python_helpers.py

```diff
@@ -123,6 +123,120 @@ def pybuild_wavefunction(mol, basis=None):
+    # collect the wavefunction's variables in a dictionary indexed by varaible type
+    # some of the data types have to be made numpy-friendly first
+    wfn_data = {
+        'molecule' : wfn.molecule().to_dict(),
+        'matrix' : {
+            'Ca'       : wfn.Ca().to_array()       if wfn.Ca()       else None,
+            'Cb'       : wfn.Cb().to_array()       if wfn.Cb()       else None,
+            'Da'       : wfn.Da().to_array()       if wfn.Da()       else None,
+            'Db'       : wfn.Db().to_array()       if wfn.Db()       else None,
+            'Fa'       : wfn.Fa().to_array()       if wfn.Fa()       else None,
+            'Fb'       : wfn.Fb().to_array()       if wfn.Fb()       else None,
+            'H'        : wfn.H().to_array()        if wfn.H()        else None,
+            'S'        : wfn.S().to_array()        if wfn.Ca()       else None,
+            'X'        : wfn.X().to_array()        if wfn.X()        else None,
+            'aotoso'   : wfn.aotoso().to_array()   if wfn.aotoso()   else None,
+            'gradient' : wfn.gradient().to_array() if wfn.gradient() else None,
+            'hessian'  : wfn.hessian().to_array()  if wfn.hessian()  else None
+            },
+        'vector' : {
+            'epsilon_a'            : wfn.epsilon_a().to_array()   if wfn.epsilon_a() else None,
+            'epsilon_b'            : wfn.epsilon_b().to_array()   if wfn.epsilon_b() else None,
+            'frequencies'          : wfn.frequencies().to_array() if wfn.frequencies() else None
+            #'atomic_point_charges' : wfn.atomic_point_charges().to_array() if wfn.atomic_point_charges() else None,
+            #'esp_at_nuclei'        : wfn.esp_at_nuclei().to_array() if wfn.esp_at_nuclei() else None,
+            },
+        'dimension' : {
+            'doccpi'   : wfn.doccpi().to_tuple(),
+            'frzcpi'   : wfn.frzcpi().to_tuple(),  
+            'frzvpi'   : wfn.frzvpi().to_tuple(),  
+            'nalphapi' : wfn.nalphapi().to_tuple(),
+            'nbetapi'  : wfn.nbetapi().to_tuple(), 
+            'nmopi'    : wfn.nmopi().to_tuple(),   
+            'nsopi'    : wfn.nsopi().to_tuple(),   
+            'soccpi'   : wfn.soccpi().to_tuple()  
+            },
+        'int' : { 
+            'nalpha' : wfn.nalpha(),
+            'nbeta'  : wfn.nbeta(),
+            'nfrzc'  : wfn.nfrzc(),
+            'nirrep' : wfn.nirrep(),
+            'nmo'    : wfn.nmo(),
+            'nso'    : wfn.nso()
+            },
+        'string' : {
+            'name' : wfn.name(),
+            'basisname' : wfn.basisset().name()
+            },
+        'boolean' : {
+            'PCM_enabled'   : wfn.PCM_enabled(),
+            'same_a_b_dens' : wfn.same_a_b_dens(),
+            'same_a_b_orbs' : wfn.same_a_b_orbs(),
+            'basispuream' : wfn.basisset().has_puream()
+            },
+        'float' : {
+            'energy' : wfn.energy(),
+            #'efzc' : wfn.efzc()
+            }
```

Comment:
Need to add variables/arrays here:

```
"variables": wfn.variables()
"arrays": {k: v.to_array() for k, v in wfn.arrays().items()}
```

The can be set by `set_variable/array`.

Reply:
Will do. That's a nice way of iterating through dictionaries.

By the way, I noticed that the wavefunction's `arrays` is always empty. This was true for both different methods (scf, mp2, ccsd, b3lyp) and different kinds of calculations (energy, gradient, hessian). Is that a bug, or have a just not encountered a calculation that actually uses it?

Reply:
It is not very common, but it does happen in a number of places where arrays are set. This is also behavior that we are trying to encourage to give users more programmatic access to internal quantities so it should certainly trend in that direction.